### ISSUE TITLE
Is there a way to redeclare an outer element as a regular one?

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#1414]
**Reported by eshmoylova on 6 Feb 2014 21:08 UTC**
If a replaceable element is declared as outer or inner I do not see a way to redeclare it as a plain element. In Section 7.3 Redeclaration it says:
  The general rule is that if no prefix within one of the following groups is present in the new declaration the old prefixes of that kind are preserved.
The inner, outer prefixes are then listed below. So I can redeclare an inner element as an outer one or the other way around. But how can I cancel the outer-ness or inner-ness of an element?

My motivation for it stems from the MSL Media and Fluid library. Different media or fluids are defined as packages with common functions and constants. The classes that use the the fluids all have a declaration of the form:
```
replaceable package Medium = SomePackage;
```
Then when instances of the classes are defined to create an example for a specific medium the redeclarations of the form:
```
redeclare package Medium = SomeOtherMedium
```
need to be passed to each and every class as it is being instantiated. It is very inconvenient to use.

An easier way would be to declare Medium as outer in the classes:
```
outer SomeMedium Medium;
```
Then when the classes are being used to make an example or a model we do not need to pass the redeclaration to every class, but can only declare one inner Medium that suits our needs. It would make the Media and Fluid library much nicer to use.

The only hesitation is how to make sure that the models that users have would still work with the new library. Right now all such models would have the above redeclaration. If that redeclaration would also remove the outer-ness of Medium, we would guarantee that all user models that work would continue to work without any alterations.

Maybe we need to introduce some other prefix that is just like outer except it can be redeclared as non-outer?

Is this an MCP-worthy idea?

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/1414

**Comment by modelica-trac-importer** [comment]
**Comment by choeger on 7 Feb 2014 12:08 UTC**
I really like the idea of using dynamic scoping for the media library. 

What I do not understand is, if it would be possible to define the medium as an outer component, then why would you need to redeclare it (and where?). If it really only is for backwards compatibility I would rather not change the language spec. Adding a feature that (ideally) is only useful a limited time seems like a bad idea.

As a second note: If the library changes, I (as a user) would never expect my models to work with the new version of the library. I understand that users have to invest a lot of effort to keep their models in sync with newer versions, but to mimize this effort we should offer long-term stable libs, instead of keeping every old model working forever.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 7 Feb 2014 13:54 UTC**
Replying to [ticket:1414 eshmoylova]:
> If a replaceable element is declared as outer or inner I do not see a way to redeclare it as a plain element. In Section 7.3 Redeclaration it says:
>   The general rule is that if no prefix within one of the following groups is present in the new declaration the old prefixes of that kind are preserved.
> The inner, outer prefixes are then listed below. So I can redeclare an inner element as an outer one or the other way around. 

As far as I recall you cannot have inner/outer as prefixes in modifiers (due to the grammar). 
So inner/outer are just part of that list so that you can redeclare an inner element and preserve inner. (Used for world-objects).

Adding outer would also cause issues with existing modifiers, and input connectors. 
Adding inner would to me usually indicate a bad model where you originally had "outer XYZ" and then at a higher level "XYZ" without inner/outer (sort of shadowing).
A redeclaring changing between outer and inner could be ok if directly inherited; but would look a bit odd.

For the Media-handling it does not seem like the best solution, even if it could work for 90% of the cases, since you sometimes have different Media and you would need additional mechanisms to handle that. There are solutions that avoid that; and they should be worked out in more detail.

**Comment by modelica-trac-importer** [comment]
**Comment by mtiller on 7 Feb 2014 15:01 UTC**
I'd like to add some comments here.  Frankly, this is the wrong approach.  I completely agree that the way media is handled can be improved greatly.  But for fluid modeling an essential component of the design should be that information about the medium is propagated **through connections** (just as the medium is).  The problem with both dynamic scoping and replaceable packages is that they necessitate specification of media information from the top down which is awkward at best.  While using dynamic scoping might reduce the amount "work" in propagating media information, it is still top-down.

Also, it is important to understand that many fluid devices (e.g., heat exchangers) have two working fluids and they may be combined into subsystems with other working fluids as well.  Propagation with dynamic scoping would be particularly problematic because of trying to make sure all the outer media declarations were picking up the correct inner media.  I just don't see that as a practical solution.

Interestingly, I think a key component to resolving this is promoting functions to first class values in Modelica.  I already wrote up an [extensive discussion of this issue](https://trac.modelica.org/Modelica/wiki/Documents/FunctionInstantiation) previously.  My thinking has changed a little since then, but I still think that functions as first class values is an essential part of the solution.  Even if we had functions as first class values, we would still face an issue with redundant equations to occurred as a result of connection loops.  But fortunately, we already have the `Connections` package to address that.

In short, I'm very much in favor of cleaning up the fluid stuff to make it simpler for the user and simpler for the compiler writer.  But I think we cannot overlook essential aspects of the problem domain when trying to craft a solution.

**Comment by modelica-trac-importer** [comment]
**Comment by eshmoylova on 7 Feb 2014 15:26 UTC**
Right, I missed the fact that there can be several different media in the same model. Just from a user point of view, I get dizzy from all the redeclare statements that need to be passed in order to make a model. I have developer's laziness: cannot stand to write explicitly the same thing over and over again. (I don't even need to explain to you how much trouble it causes for a Modelica tool implementer.)

Mike, I'll check the other discussion, and close this ticket.

**Comment by modelica-trac-importer** [comment]
**Modified by eshmoylova on 7 Feb 2014 15:26 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 7 Feb 2014 15:30 UTC**

**Comment by modelica-trac-importer** [comment]
**Modified by dietmarw on 7 Feb 2014 15:30 UTC**
